File: old_versions/v2/tracker/services/dashboard_service.py
from datetime import datetime, time, date, timedelta
from dateutil.relativedelta import relativedelta

def get_time_period(period_str, today_date):
    match period_str:
        case 'day':
            start = datetime.combine(today_date, time.min)
            end = datetime.combine(today_date, time.max)
            chart_labels = [f"{h:02d}:00" for h in range(24)]
        case 'week':
            start_of_week = today_date - timedelta(days=today_date.weekday())
            end_of_week = start_of_week + timedelta(days=6)
            start = datetime.combine(start_of_week, time.min)
            end = datetime.combine(end_of_week, time.max)
            chart_labels = ['Mo', 'Di', 'Mi', 'Do', 'Fr', 'Sa', 'So']
        case 'month':
            start_of_month = today_date.replace(day=1)
            end_of_month = (start_of_month + relativedelta(months=1)) - timedelta(days=1)
            start = datetime.combine(start_of_month, time.min)
            end = datetime.combine(end_of_month, time.max)
            chart_labels = [str(d) for d in range(1, end_of_month.day + 1)]
        case 'year':
            start_of_year = today_date.replace(month=1, day=1)
            end_of_year = today_date.replace(month=12, day=31)
            start = datetime.combine(start_of_year, time.min)
            end = datetime.combine(end_of_year, time.max)
            chart_labels = ['Jan', 'Feb', 'Mär', 'Apr', 'Mai', 'Jun', 'Jul', 'Aug', 'Sep', 'Okt', 'Nov', 'Dez']
        case _:
            return None, None, None
    return start, end, chart_labels

def calculate_bmr(profile):
    if not (profile and profile.gender and profile.age and profile.height_cm and profile.weight_kg):
        return 0, None
    
    bmr_val = 0
    if profile.gender.lower() == 'female':
        bmr_val = 655.1 + (9.6 * profile.weight_kg) + (1.8 * profile.height_cm) - (4.7 * profile.age)
    elif profile.gender.lower() == 'male':
        bmr_val = 66.47 + (13.7 * profile.weight_kg) + (5 * profile.height_cm) - (6.8 * profile.age)
    
    return round(bmr_val), profile.tracking_start_date

def process_logs(logs, period, num_labels, model_class):
    calories_list = [0] * num_labels
    details_list = [[] for _ in range(num_labels)]
    
    def get_interval_index(log_date_local, period_str):
        if period_str == 'day': return log_date_local.hour
        if period_str == 'week': return log_date_local.weekday()
        if period_str == 'month': return log_date_local.day - 1
        if period_str == 'year': return log_date_local.month - 1
        return -1

    for log in logs:
        is_consumption = model_class.__name__ == 'ConsumptionLog'
        calories = (log.amount_g / 100 * log.food.calories_per_100g) if is_consumption else (log.duration_min / 60 * log.exercise_type.calories_per_hour)
        name = log.food.name if is_consumption else log.exercise_type.name
        
        idx = get_interval_index(log.log_date.astimezone(), period)
        if 0 <= idx < num_labels:
            calories_list[idx] += calories
            details_list[idx].append(f"{name}: {round(calories)} kcal")
    
    return [round(c) for c in calories_list], details_list

def add_bmr_to_calories_out(calories_out_list, daily_bmr, period_str, start_date_dt, today_date, tracking_start):
    if not (daily_bmr > 0 and tracking_start):
        return calories_out_list

    match period_str:
        case 'day':
            limit_index = datetime.now().hour
            bmr_per_interval = daily_bmr / 24
            day_date = start_date_dt.date()
            if day_date >= tracking_start:
                for i in range(len(calories_out_list)):
                    if i <= limit_index:
                        calories_out_list[i] += bmr_per_interval
        case 'week' | 'month':
            limit_index = today_date.weekday() if period_str == 'week' else today_date.day - 1
            for i in range(len(calories_out_list)):
                day_date = start_date_dt.date() + timedelta(days=i)
                if i <= limit_index and day_date >= tracking_start:
                    calories_out_list[i] += daily_bmr
        case 'year':
            limit_month_index = today_date.month - 1
            for i in range(len(calories_out_list)):
                month_start = date(today_date.year, i + 1, 1)
                if tracking_start and month_start < tracking_start.replace(day=1):
                    continue
                
                if i < limit_month_index:
                    days_in_month = (month_start + relativedelta(months=1) - timedelta(days=1)).day
                    calories_out_list[i] += daily_bmr * days_in_month
                elif i == limit_month_index:
                    start_day = max(1, tracking_start.day if tracking_start and tracking_start.month == i + 1 else 1)
                    days_to_add = today_date.day - start_day + 1
                    if days_to_add > 0:
                        calories_out_list[i] += daily_bmr * days_to_add
    
    return [round(c) for c in calories_out_list]

File: V6_fastAPI_frontend_improved/tracker/static/js/services/Dashboard_Service.js
// This service provides pure functions for calculating dashboard data.
// It relies on date-fns for robust date manipulation. A real project
// should include it via npm/yarn, but here we assume it's globally available.
// A simple date-fns build can be included via: <script src="https://cdn.jsdelivr.net/npm/date-fns@2.28.0/cdn.min.js"></script>

function _get_time_period(period_string) {
    const today = new Date();
    let start, end, labels = [];

    switch (period_string) {
        case 'day':
            start = new Date(today.setHours(0, 0, 0, 0));
            end = new Date(today.setHours(23, 59, 59, 999));
            labels = ['Heute'];
            break;
        case 'week':
            // Assuming week starts on Monday
            const day_of_week = today.getDay(); // 0=Sun, 1=Mon, ..., 6=Sat
            const diff = today.getDate() - day_of_week + (day_of_week === 0 ? -6 : 1);
            start = new Date(new Date(today.setDate(diff)).setHours(0, 0, 0, 0));
            end = new Date(new Date(start).setDate(start.getDate() + 6));
            end.setHours(23, 59, 59, 999);
            labels = ['Mo', 'Di', 'Mi', 'Do', 'Fr', 'Sa', 'So'];
            break;
        case 'month':
            start = new Date(today.getFullYear(), today.getMonth(), 1);
            end = new Date(today.getFullYear(), today.getMonth() + 1, 0, 23, 59, 59, 999);
            labels = Array.from({ length: end.getDate() }, (_, i) => i + 1);
            break;
        case 'year':
            start = new Date(today.getFullYear(), 0, 1);
            end = new Date(today.getFullYear(), 11, 31, 23, 59, 59, 999);
            labels = ['Jan', 'Feb', 'Mär', 'Apr', 'Mai', 'Jun', 'Jul', 'Aug', 'Sep', 'Okt', 'Nov', 'Dez'];
            break;
        default:
            return { start: null, end: null, labels: [] };
    }
    return { start, end, labels };
}

function _calculate_BMR(profile) {
    if (!profile || !profile.gender || !profile.age || !profile.height_cm || !profile.weight_kg) {
        return 0;
    }
    if (profile.gender.toLowerCase() === 'female') {
        return 655.1 + (9.6 * profile.weight_kg) + (1.8 * profile.height_cm) - (4.7 * profile.age);
    }
    if (profile.gender.toLowerCase() === 'male') {
        return 66.47 + (13.7 * profile.weight_kg) + (5 * profile.height_cm) - (6.8 * profile.age);
    }
    return 0;
}

function _process_logs(logs, period, numLabels, { nameKey, valueKey }) {
    const values = Array(numLabels).fill(0);
    const details = Array.from({ length: numLabels }, () => []);
    
    const get_interval_index = (logDate) => {
        if (period === 'day') return 0;
        if (period === 'week') return (logDate.getDay() + 6) % 7; // Monday = 0
        if (period === 'month') return logDate.getDate() - 1;
        if (period === 'year') return logDate.getMonth();
        return -1;
    };

    logs.forEach(log => {
        const logDate = new Date(log.log_date);
        const idx = get_interval_index(logDate);
        if (idx >= 0 && idx < numLabels) {
            values[idx] += log.calories;
            details[idx].push(`${log[nameKey]}: ${Math.round(log.calories)} kcal`);
        }
    });
    return { values: values.map(Math.round), details };
}

export const Dashboard_Service = {
    get_dashboard_data: (period, all_data) => {
        const { start, end, labels } = _get_time_period(period);
        if (!start) return {}; // Invalid period

        // **KORREKTUR: Zugriff auf normalisierte Keys**
        const { consumption_logs, activity_logs, user_profile } = all_data;

        const relevant_consumptions = consumption_logs.filter(log => new Date(log.log_date) >= start && new Date(log.log_date) <= end);
        const relevant_activities = activity_logs.filter(log => new Date(log.log_date) >= start && new Date(log.log_date) <= end);

        const { values: calories_in, details: details_in } = _process_logs(relevant_consumptions, period, labels.length, { nameKey: 'food_name', valueKey: 'calories' });
        const { values: calories_out_active, details: details_out } = _process_logs(relevant_activities, period, labels.length, { nameKey: 'exercise_name', valueKey: 'calories' });
        
        const daily_BMR = _calculate_BMR(user_profile);
        const calories_out_bmr = labels.map(() => Math.round(daily_BMR)); // Simplified for daily BMR, can be expanded

        const total_in = calories_in.reduce((a, b) => a + b, 0);
        const total_out_active = calories_out_active.reduce((a, b) => a + b, 0);
        // This BMR calculation is simplified and should be made more precise for month/year views
        const total_out_bmr = Math.round(daily_BMR * (period === 'day' ? 1 : (period === 'week' ? 7 : 30)));
        const total_out = total_out_active + total_out_bmr;
        const balance = total_in - total_out;

        return {
            labels,
            calories_in,
            calories_out_active,
            calories_out_bmr,
            details_in,
            details_out,
            total_in,
            total_out,
            balance,
            balance_goal: user_profile?.balance_goal_kcal || 0
        };
    }
};
